book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+
+### Open vs Closed Object Types
+
+TypeScript, by default, treats all objects as _open_. At any time, it expects that other properties might be present on objects.
+
+Other languages, like Flow, treat objects as _closed_ by default. Flow is Meta's internal type system, and by default requires objects to be exact (their term for 'closed').
+
+```js
+function method(obj: { foo: string }) {
+  /* ... */
+}
+
+method({ foo: "test", bar: 42 }); // Error!
+```
+
+You can opt in to open (or inexact) objects in Flow with a `...` syntax:
+
+```js
+function method(obj: { foo: string, ... }) {
+  /* ... */
+}
+
+method({ foo: "test", bar: 42 }); // No more error!
+```
+
+But Flow recommends you use closed objects by default. They think that, especially when working with spread operators, it's better to err on the side of caution.
+
+### Why Does TypeScript Treat Objects As Open?
+
+Open objects more closely reflect how JavaScript actually works. Any type system for JavaScript - a very dynamic language - has to be relatively cautious about how 'safe' it can truly be.
+
+So, TypeScript's decision to treat objects as open by default is a reflection of the language it's trying to type. It also more closely reflects how objects work in other languages.
+
+The issue is that the excess properties warning can often make you think TypeScript uses closed objects.
+
+But really, the excess properties warning is more like a courtesy. It's only used in cases where the object can't be modified elsewhere.
+
+## Object Keys Are Loosely Typed
+
+A consequence of TypeScript having open object types is that iterating over the keys of an object can be frustrating.
+
+In JavaScript, calling `Object.keys` with an object will return an array of strings representing the keys.
+
+```ts twoslash
+const yetiSeason = {
+  title: "Yeti Season",
+  artist: "El Michels Affair",
+  releaseYear: 2021,
+};
+
+const keys = Object.keys(yetiSeason);
+//    ^?
+```
+
+In theory, you can then use those keys to access the values of the object:
+
+```ts twoslash
+// @errors: 7053
+const yetiSeason = {
+  title: "Yeti Season",
+  artist: "El Michels Affair",
+  releaseYear: 2021,
+};
+
+const keys = Object.keys(yetiSeason);
+// ---cut---
+keys.forEach((key) => {
+  console.log(yetiSeason[key]); // Red squiggly line under key
+});
+```
+
+But we're getting an error. TypeScript is telling us that we can't use `string` to access the properties of `yetiSeason`.
+
+The only way this would work would be if `key` was typed as `'title' | 'artist' | 'releaseYear'`. In other words, as `keyof typeof yetiSeason`. But it's not - it's typed as `string`.
+
+The reason for this comes back to `Object.keys` - it returns `string[]`, not `(keyof typeof obj)[]`.
+
+```ts twoslash
+// @errors: 2304
+const keys = Object.keys(yetiSeason);
+//     ^?
+```
+
+By the way, the same behavior happens with `for ... in` loops:
```

Comment:
funny thing is that TS doesn't complain about this at all ([TS playground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAngUygSwMoIIYXDAvDAbwCgYYUoAbBALhgCIBNJZGdLcOgGhJgwCcU0WnQCiFGAFlkwABYIKEGAEEAZiozI+XHn3mYICJv1oAmAAwmAjNwC+AbiJEqsANYI4tN3BAqycAA4IPvDMbNhgRCogfDAAFF4wyGAhKGHgAJSEPKCQIFQAdBQgAOaxiKn64ADaXgC66Q42QA)):
```ts
let key: keyof typeof yetiSeason
for (key in yetiSeason) {
  console.log(yetiSeason[key]);
}
```

it is an intentional type hole 🤷‍♂️ I wouldn't recommend relying on this though